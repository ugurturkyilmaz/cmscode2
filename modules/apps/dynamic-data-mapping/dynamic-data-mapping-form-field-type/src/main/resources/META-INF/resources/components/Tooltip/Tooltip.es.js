/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

/* eslint-disable react/no-string-refs */

import '../FormPortal/FormPortal.es';

import 'clay-icon';
import {ALIGN_POSITIONS, align} from 'frontend-js-web';
import Component from 'metal-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './Tooltip.soy';

const POSITIONS = ['top', 'left', 'right', 'bottom'];

class Tooltip extends Component {
	created() {
		this._handleTooltipHovered = this._handleTooltipHovered.bind(this);
		this._handleTooltipRendered = this._handleTooltipRendered.bind(this);
	}

	_handleTooltipHovered() {
		this.showContent = true;
	}

	_handleTooltipLeaved() {
		this.showContent = false;
	}

	_handleTooltipRendered() {
		const {tooltipSource, tooltipTarget} = this.refs;
		const {element} = tooltipSource;
		const suggestedPosition = align(
			element,
			tooltipTarget,
			ALIGN_POSITIONS.Right
		);

		this.position = POSITIONS[suggestedPosition];
	}
}

Soy.register(Tooltip, templates);

Tooltip.STATE = {

	/**
	 * @default undefined
	 * @instance
	 * @memberof Tooltip
	 * @type {?(string|undefined)}
	 */

	icon: Config.string().required(),

	/**
	 * @default undefined
	 * @instance
	 * @memberof Tooltip
	 * @type {?(number|undefined)}
	 */

	position: Config.string(),

	/**
	 * @default undefined
	 * @instance
	 * @memberof Tooltip
	 * @type {?(string|undefined)}
	 */

	showContent: Config.bool(),

	/**
	 * @default undefined
	 * @instance
	 * @memberof Tooltip
	 * @type {?(string|undefined)}
	 */

	spritemap: Config.string().required(),

	/**
	 * @default undefined
	 * @instance
	 * @memberof Tooltip
	 * @type {?(string|undefined)}
	 */

	text: Config.string().required(),
};

export default Tooltip;
