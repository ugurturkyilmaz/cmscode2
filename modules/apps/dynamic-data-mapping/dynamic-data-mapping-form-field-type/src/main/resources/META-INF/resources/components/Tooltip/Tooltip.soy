{namespace Tooltip}

{template .render}
	{@param icon: string}
	{@param? _handleTooltipHovered: any}
	{@param? _handleTooltipLeaved: any}
	{@param? _handleTooltipRendered: any}
	{@param? position: string}
	{@param? showContent: bool}
	{@param? spritemap: string}
	{@param? text: string}

	<div class="ddm-tooltip">
		<div data-onmouseleave="{$_handleTooltipLeaved}" data-onmouseover="{$_handleTooltipHovered}" ref="tooltipTarget">
			<svg
				aria-hidden="true"
				class="lexicon-icon lexicon-icon-{$icon}"
			>
				<use xlink:href="{$spritemap}#{$icon}" />
			</svg>
		</div>

		{let $content kind="html"}
			{call Tooltip.content}
				{param position: $position /}
				{param text: $text /}
			{/call}
		{/let}

		{if $showContent}
			{call FormPortal.render}
				{param contentRenderer kind="html"}
					{$content}
				{/param}
				{param ref: 'tooltipSource' /}
				{param portalElement: '#clay_dropdown_portal' /}
				{param events: [
					'rendered': $_handleTooltipRendered
				] /}
			{/call}
		{/if}

	</div>
{/template}

{template .content}
	{@param? position: string}
	{@param? text: string}

	<div class="clay-tooltip-{$position} ddm-tooltip-container tooltip" role="tooltip">
		<div class="arrow"></div>
		<div class="tooltip-inner">
			<div class="ddm-tooltip-text">{$text}</div>
		</div>
	</div>
{/template}