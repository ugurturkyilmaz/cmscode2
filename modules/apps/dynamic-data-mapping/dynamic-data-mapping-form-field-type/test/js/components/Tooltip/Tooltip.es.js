/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import userEvent from '@testing-library/user-event';

import Tooltip from '../../../src/main/resources/META-INF/resources/components/Tooltip/Tooltip.es';

let component;
const spritemap = 'icons.svg';

describe('Field Tooltip', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders the default markup', () => {
		component = new Tooltip({
			icon: 'question-circle-full',
			spritemap,
			text: 'This is a tooltip information about this component',
		});

		expect(component).toMatchSnapshot();
	});

	it('updates the tooltip visible state when the mouse is over the tooltip target', () => {
		component = new Tooltip({
			icon: 'question-circle-full',
			spritemap,
			text: 'This is a tooltip information about this component',
		});

		jest.runAllTimers();

		const {tooltipTarget} = component.refs;

		userEvent.hover(tooltipTarget);

		expect(component.showContent).toBe(true);
		expect(component).toMatchSnapshot();
	});

	it('updates the tooltip visible state when the mouse leaved the tooltip target', () => {
		component = new Tooltip({
			icon: 'question-circle-full',
			spritemap,
			text: 'This is a tooltip information about this component',
		});

		component._handleTooltipHovered();
		component._handleTooltipLeaved();

		expect(component.showContent).toBe(false);
		expect(component).toMatchSnapshot();
	});

	it('updates the tooltip visible state when the mouse is over the tooltip target', () => {
		component = new Tooltip({
			icon: 'question-circle-full',
			spritemap,
			text: 'This is a tooltip information about this component',
		});

		jest.runAllTimers();

		component.refs.tooltipSource = {
			element: document.createElement('div'),
		};

		component._handleTooltipHovered();
		component._handleTooltipRendered();

		expect(component.showContent).toBe(true);
		expect(component).toMatchSnapshot();
	});
});
