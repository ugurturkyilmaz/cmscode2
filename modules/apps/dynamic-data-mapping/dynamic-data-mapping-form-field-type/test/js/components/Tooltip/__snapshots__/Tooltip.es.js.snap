// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Tooltip should render the default markup 1`] = `
<div>
  <div class="ddm-tooltip">
    <div ref="tooltipTarget"><svg aria-hidden="true" class="lexicon-icon lexicon-icon-question-circle-full">
        <use xlink:href="icons.svg#question-circle-full"></use>
      </svg></div>
  </div>
</div>
`;

exports[`Field Tooltip should update the tooltip visible state when the mouse is over the tooltip target 1`] = `
<div>
  <div class="ddm-tooltip">
    <div ref="tooltipTarget"><svg aria-hidden="true" class="lexicon-icon lexicon-icon-question-circle-full">
        <use xlink:href="icons.svg#question-circle-full"></use>
      </svg></div>
  </div>
</div>
`;

exports[`Field Tooltip should update the tooltip visible state when the mouse is over the tooltip target 2`] = `
<div class="ddm-tooltip">
  <div ref="tooltipTarget"><svg aria-hidden="true" class="lexicon-icon lexicon-icon-question-circle-full">
      <use xlink:href="icons.svg#question-circle-full"></use>
    </svg></div>
</div>
`;

exports[`Field Tooltip should update the tooltip visible state when the mouse leaved the tooltip target 1`] = `
<div>
  <div class="ddm-tooltip">
    <div ref="tooltipTarget"><svg aria-hidden="true" class="lexicon-icon lexicon-icon-question-circle-full">
        <use xlink:href="icons.svg#question-circle-full"></use>
      </svg></div>
  </div>
</div>
`;
